core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFormatterUtil.java

```diff
@@ -201,6 +254,90 @@ static ExprValue getFormattedTime(ExprValue timeExpr, ExprValue formatExpr) {
     return getFormattedString(formatExpr, TIME_HANDLERS, time);
   }
 
+  private static boolean canGetDate(TemporalAccessor ta) {
+    if (!ta.isSupported(ChronoField.YEAR)
+        || !ta.isSupported(ChronoField.MONTH_OF_YEAR)
+        || !ta.isSupported(ChronoField.DAY_OF_MONTH)) {
+      return false;
+    }
+    return true;
+  }
+
+  private static boolean canGetTime(TemporalAccessor ta) {
+    if (!ta.isSupported(ChronoField.HOUR_OF_DAY)
+        || !ta.isSupported(ChronoField.MINUTE_OF_HOUR)
+        || !ta.isSupported(ChronoField.SECOND_OF_MINUTE)) {
+      return false;
+    }
+    return true;
+  }
+
+  static ExprValue parseStringWithDateOrTime(FunctionProperties fp,
+                                             ExprValue datetimeStringExpr,
+                                             ExprValue formatExpr) {
+
+    //Replace patterns with % for Java DateTimeFormatter
+    StringBuffer cleanFormat = getCleanFormat(formatExpr);
+    final Matcher matcher = pattern.matcher(cleanFormat.toString());
+    final StringBuffer format = new StringBuffer();
+
+    while (matcher.find()) {
+      matcher.appendReplacement(format,
+          STR_TO_DATE_FORMATS.getOrDefault(matcher.group(),
+              String.format("'%s'", matcher.group().replaceFirst(MOD_LITERAL, ""))));
+    }
+    matcher.appendTail(format);
+
+    TemporalAccessor taWithMissingFields;
+    //Return NULL for invalid parse in string to align with MySQL
+    try {
+      //Get Temporal Accessor to initially parse string without default values
+      taWithMissingFields = new DateTimeFormatterBuilder()
+          .appendPattern(format.toString())
+          .toFormatter().withResolverStyle(ResolverStyle.STRICT)
+          .parseUnresolved(datetimeStringExpr.stringValue(), new ParsePosition(0));
+      if (taWithMissingFields == null) {
+        throw new DateTimeException("Input string could not be parsed properly.");
+      }
+      if (!canGetDate(taWithMissingFields) && !canGetTime(taWithMissingFields)) {
+        throw new DateTimeException("Not enough data to build a valid Date, Time, or Datetime.");
```

Comment:
There is no decision yet in #1423, but maybe return `NULL` in all error cases?

Reply:
I don't see UTs with `assertThrows`, is it covered by tests?

Reply:
No, the actual exception being thrown is not checked by tests. Since according to MySQL a `NULL` should be returned if an error is found at this stage, there are tests to check if `NULL` is returned instead. The thrown exception causes `ExprNullValue.of()` to be returned below instead.

Reply:
Oh I see, thanks